mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala

```diff
@@ -208,20 +239,61 @@ class LinearRegression(override val uid: String)
 class LinearRegressionModel private[ml] (
     override val uid: String,
     val weights: Vector,
-    val intercept: Double)
+    val intercept: Double,
+    val trainingSummary: LinearRegressionTrainingResults)
   extends RegressionModel[Vector, LinearRegressionModel]
   with LinearRegressionParams {
 
+  /**
+   * Evaluates the model on a test-set.
+   * @param testset Test dataset to evaluate model on.
+   * @return
+   */
+  def evaluate(testset: DataFrame): LinearRegressionResults = {
+    val predictionAndObservations = testset
+      .select($(labelCol), $(featuresCol))
+      .map { case Row(label: Double, features: Vector) => (predict(features), label) }
+
+    new LinearRegressionResults(predictionAndObservations)
+  }
+
   override protected def predict(features: Vector): Double = {
     dot(features, weights) + intercept
   }
 
   override def copy(extra: ParamMap): LinearRegressionModel = {
-    copyValues(new LinearRegressionModel(uid, weights, intercept), extra)
+    copyValues(new LinearRegressionModel(uid, weights, intercept, trainingSummary), extra)
   }
 }
 
 /**
+ * :: Experimental ::
+ * Linear regression training results.
+ * @param predictionAndObservations dataframe with columns predictions (0) and observations(0).
+ * @param totalIterations total number of training iterations.
+ * @param objectiveTrace objective function value at each iteration.
+ */
+@Experimental
+class LinearRegressionTrainingResults(
+    @transient predictionAndObservations: RDD[(Double, Double)],
+    val totalIterations: Int,
+    val objectiveTrace: Array[Double])
+  extends LinearRegressionResults(predictionAndObservations)
+
+/**
+ * :: Experimental ::
+ * Linear regression results evaluated on a dataset.
+ * @param predictionAndObservations an RDD of (prediction, observation) pairs.
+ */
+@Experimental
+class LinearRegressionResults(@transient val predictionAndObservations: RDD[(Double, Double)])
+  extends RegressionMetrics(predictionAndObservations) {
```

Comment:
I like the code-sharing motivation behind extending RegressionMetrics, but we should not include this dependency between the ml and mllib packages.  We can use RegressionMetrics internally, but we should create a copy of each of the methods in this class.  (Or those methods could be in a private abstraction which can be shared for later regression models.)